### PR TITLE
calib: axis swapping in `accorth`

### PR DESCRIPTION
**Pull request by niewim19**
## Description

Addition of axis swapping/inversion solution in IMU initial rotation estimation. From now on measurements of IMU are not only rotated using quaternion but also their order is switched and inverted if necessary. 

## Motivation and Context

Estimating IMU rotation only with quaternion has a room for an error if that rotation is big/complex (not along one axis). Instead of using only quaternion it may be easier to firstly switch axis order and/or invert some of them and then calculate the quaternion. 

This way big rotations are not included in the quaternion so it more correctly estimates the missalignment of the axis. This also makes the calibration procedure more robust i.e. differences between consecutive calibration procedures in calibtool are smaller

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: zturn drone.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: libs/calib/calib.h [hunk 1]
```diff
@@ -60,9 +66,11 @@ typedef struct {
 		} magiron;
 
 		struct {
-			matrix_t ortho;  /* 3x3 nonorthogonality parameters matrix */
-			matrix_t offset; /* 3x1 measurement offset matrix */
-			quat_t frameQ;   /* initial rotation quaternion of accelerometer in relation to body frame */
+			matrix_t ortho;      /* 3x3 nonorthogonality parameters matrix */
+			matrix_t offset;     /* 3x1 measurement offset matrix */
+			quat_t frameQ;       /* initial rotation quaternion of accelerometer in relation to body frame */
+			accSwap_t swapOrder; /* axis swap order: e.g 312 means that [x,y,z] will be read as [z, x, y] */
```
Comment (PNieck) [inline libs/calib/calib.h:72]: Misleading comment, which suggest that you are using tree digit numbers to remember swap order, when in reality program uses enum values

File: calib/accorth.c [hunk 2]
```diff
@@ -230,8 +238,129 @@ static int accorth_accRot(matrix_t *S, matrix_t *H, quat_t *rotQuat)
...
-	/* Calculate rotation */
-	quat_frameRot(&bodyZ, &bodyY, &nedZ, &nedY, &q, &idenQ);
+	/*
+	 * Quaternion imaginary part minimization via axis switching/inversion
+	 *
+	 * Direct IMU rotation quaternion calculation ends up with high error if the total rotation angle is big.
+	 * If rotation is somewhat along the IMU axis and greater than pi/4 it is better to switch axis order and/or negate some axis.
+	 * This way rotation quaternion only handles missalignment of axis and not the whole rotation.
+	 *
+	 * Axis switching search is done by bruteforcing all combinations, 24 total.
```
Comment (PNieck) [inline calib/accorth.c:248]: If my calculations are correct there are 48 combinations. You can order arises in 6 ways and then negate them or not in 8 ways.

**Review by PNieck — CHANGES_REQUESTED** [review]

File: calib/accorth.c (hunk 2 above)
Comment (niewim19) [reply calib/accorth.c:248]: Yes, there are 48 arbitrary combinations, but only half of them follow the right-hand rule.

Comment (PNieck) [reply calib/accorth.c:248]: Yes, you are right

Comment (PNieck) [reply calib/accorth.c:248]:
```suggestion
	 * Axis switching search is done by brute forcing all right-hand combinations, 24 total.
```

**Review by PNieck — CHANGES_REQUESTED** [review]

File: libs/calib/calib.h [hunk 3]
```diff
@@ -34,10 +34,13 @@
 #define HARDCAL_COLSPAN 1
 
 #define ACCORTH_TAG        "accorth"
-#define ACCORTH_PARAMS     16
+#define ACCORTH_PARAMS     20
 #define ACC_CHAR_ORTHO     'o'
 #define ACC_CHAR_OFFSET    'h'
 #define ACC_CHAR_QUAT      'q'
+#define ACC_CHAR_SWAP      's'
+#define ACC_CHAR_SWAP_SIGN 's'
```
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/calib/calib.h:42]: They will be in different positions in line. `ACC_CHAR_SWAP` occupies index 0, and `ACC_CHAR_SWAP_SIGN` occupies index 1.

Final form in `calib.conf`: https://github.com/phoenix-pilot/phoenix-pilot-project/blob/acd1b2fa91288173cf1933b5186f953e7a77f6ab/_projects/armv7a9-zynq7000-pilot/rootfs-overlay/etc/calib.conf#L12-L32

File: calib/accorth.c [hunk 4]
```diff
@@ -230,8 +238,129 @@ static int accorth_accRot(matrix_t *S, matrix_t *H, quat_t *rotQuat)
...
+
+					case 0:
+					default:
+						/* leave as is */
+						break;
+				}
+
+				/* Find quaternion for current axis switch and compare to current minimal */
+				quat_frameRot(&bodyZ, &bodyY, &nedZ, &nedY, &q, &idenQ);
+				imLen = sqrt(q.i * q.i + q.j * q.j + q.k * q.k);
```
[the remark this answers is not in the document]
Comment (niewim19) [reply calib/accorth.c:352]: Yeah, good idea, but as you mentioned - this is not performance-intensive code. I would rather stay with `sqrt` as to make it easy to see in the coda what is happening.

File: libs/calib/calib.h (hunk 1 above)
Comment (PNieck) [reply libs/calib/calib.h:72]: This issue is still not fixed

Comment (niewim19) [reply libs/calib/calib.h]: Thanks! Updated.

**Review by PNieck — APPROVED** [review]
LGTM

**Pull request merged**